std/socket.d

```diff
@@ -289,56 +289,96 @@ bool wouldHaveBlocked() nothrow @nogc
     assert(rec == -1 && wouldHaveBlocked());
 }
 
+// Returns true iff getnameinfo is supported on this platform.
+version(Windows) private @property bool getnameinfoSupported()
+{
+    initialize;
+    return getnameinfoPointer !is null;
+}
+// Ditto (always supported on Posix).
+version(Posix) private enum bool getnameinfoSupported = true;
 
-private immutable
+// Returns true iff getaddrinfo is supported on this platform.
+version(Windows) private @property bool getaddrinfoSupported()
 {
-    typeof(&getnameinfo) getnameinfoPointer;
-    typeof(&getaddrinfo) getaddrinfoPointer;
-    typeof(&freeaddrinfo) freeaddrinfoPointer;
+    initialize;
+    assert((getaddrinfoPointer is null) == (freeaddrinfoPointer is null));
+    return getaddrinfoPointer !is null;
 }
+// Ditto (always supported on Posix).
+version(Posix) private enum bool getaddrinfoSupported = true;
 
-shared static this() @system
+version(Windows) private
 {
-    version (Windows)
-    {
-        WSADATA wd;
+    import core.atomic;
 
-        // Winsock will still load if an older version is present.
-        // The version is just a request.
-        int val;
-        val = WSAStartup(0x2020, &wd);
-        if (val)         // Request Winsock 2.2 for IPv6.
-            throw new SocketOSException("Unable to initialize socket library", val);
+    shared typeof(&getnameinfo) getnameinfoPointer;
+    shared typeof(&getaddrinfo) getaddrinfoPointer;
+    shared typeof(&freeaddrinfo) freeaddrinfoPointer;
 
-        // These functions may not be present on older Windows versions.
-        // See the comment in InternetAddress.toHostNameString() for details.
-        auto ws2Lib = GetModuleHandleA("ws2_32.dll");
-        if (ws2Lib)
-        {
-            getnameinfoPointer = cast(typeof(getnameinfoPointer))
-                                 GetProcAddress(ws2Lib, "getnameinfo");
-            getaddrinfoPointer = cast(typeof(getaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "getaddrinfo");
-            freeaddrinfoPointer = cast(typeof(freeaddrinfoPointer))
-                                 GetProcAddress(ws2Lib, "freeaddrinfo");
-        }
+    auto our_getnameinfo(A...)(A a) @system
+    {
+        initialize;
+        return getnameinfoPointer(a);
     }
-    else version (Posix)
+
+    auto our_getaddrinfo(A...)(A a) @system
     {
-        getnameinfoPointer = &getnameinfo;
-        getaddrinfoPointer = &getaddrinfo;
-        freeaddrinfoPointer = &freeaddrinfo;
+        initialize;
+        return getaddrinfoPointer(a);
     }
-}
 
+    auto our_freeaddrinfo(A...)(A a) @system
+    {
+        initialize;
+        if (freeaddrinfoPointer)
+            // May be null if WSAStartup() failed, just do nothing
+            return;
+        return freeaddrinfoPointer(a);
+    }
 
-shared static ~this() @system nothrow @nogc
-{
-    version (Windows)
+    void initialize() @trusted nothrow
```

Comment:
If I'm reading this correctly, this function needs to be called in a lot more places now (every single bit of public code that eventually calls WSA functions, so that it calls `WSAStartup`), and I don't see this patch doing it.

Reply:
Got it, will fix. Thanks.